Objects/dictobject.c

```diff
@@ -3230,17 +3342,31 @@ dict_equal(PyDictObject *a, PyDictObject *b)
         /* can't be equal if # of entries differ */
         return 0;
     /* Same # of entries -- check all of 'em.  Exit early on any diff. */
-    for (i = 0; i < a->ma_keys->dk_nentries; i++) {
+    Py_ssize_t n_entries;
+#ifdef Py_GIL_DISABLED
+    if (_PyDict_HasSplitTable(a)) {
+        // New entries can be appended, but existing ones won't change, and
+        // our keys won't change because the dict is locked.  So capture
+        // the current number of keys at entry.
+        LOCK_KEYS(a->ma_keys);
+        n_entries = a->ma_keys->dk_nentries;
+        UNLOCK_KEYS(a->ma_keys);
+    } else
+#endif
+    {
+        n_entries = a->ma_keys->dk_nentries;
+    }
+    for (i = 0; i < n_entries; i++) {
         PyObject *key, *aval;
         Py_hash_t hash;
         if (DK_IS_UNICODE(a->ma_keys)) {
             PyDictUnicodeEntry *ep = &DK_UNICODE_ENTRIES(a->ma_keys)[i];
-            key = ep->me_key;
+            key = LOAD_SHARED_KEY(ep->me_key);
```

Comment:
We should be locking the `a` and `b` dicts, which means we should just use plain loads for `ep->me_key`.